proposals/csharp-10.0/improved-interpolated-strings.md

```diff
@@ -452,6 +452,46 @@ rented array inside it, and if one of the interpolation holes throws an exceptio
 _Answer_: No. handlers can be assigned to locals (such as `MyHandler handler = $"{MyCode()};`), and the lifetime of such handlers is unclear. Unlike foreach enumerators, where the lifetime
 is obvious and no user-defined local is created for the enumerator.
 
+### Impact on nullable reference types
+
+To minimize complexity of the implementation, we have a few limitations on how we perform nullable analysis on interpolated string handler constructors used as arguments to a method or indexer.
+In particular, we do not flow information from the constructor back through to the original slots of parameters or arguments from the original context, and we do not use constructor parameter
+types to inform generic type inference for type parameters in the containing method. An example of where this can have an impact is:
+
+```cs
+string s = """";
+C c = new C();
+c.M(s, $"""", c.ToString(), s.ToString()); // No warnings on c.ToString() or s.ToString(), as the `MaybeNull` does not flow back.
```

Comment:
I wouldn't expect `[MaybeNull]` to have any impact here (interpolated strings are no). That attribute should only be impacting the input position, not the output position. What am I missing?

Reply:
AllowNull is for inputs, MaybeNull is for outputs.

Reply:
Huh, seems odd for that to have value in non-ref parameters. That isn't an output

Reply:
Consider `Assert.NotNull`. That will use a `NotNull` attribute on a by-val parameter and throw/assert if it's `null`.

Reply:
Correct. But it's reasonable for an API to strengthen our understanding of a piece of nullable state. The implementation of a method can do all manner of tricks to verify that a piece of input is not null. Effectively elevating the state from "maybe null" to "not null".

Having `[MaybeNull]` impact the state is saying that an API can downgrade the state. Taking something the compiler believes is "not null" and making it "maybe null". Having trouble seeing how an API can make such an assertion. Basically because you passed A to me you should assume A is now possible null.

Reply:
I don't see a real reason to have it not impact things, and I don't think that this spec is the place to make the argument to change the feature in any case 😀

Reply:
Don't want my debate to be about changing the feature. Want to make sure I understand `[MaybeNull]` so I can better evaluate this change.

Reply:
Basically, in an ideal world, this feature would behave exactly as if you called all these methods in order manually, and that would mean any nullability post-condition attributes flowing back to the original expression (`MaybeNull` included). For implementation simplicity, we have avoided this.